### ISSUE TITLE
[Windows] Run pub in batch will cause unexpectedly exit

### ISSUE THREAD

**Issue by AlexV525** [dart-lang/pub#2873]
# Environment

- `pub version` or `flutter pub version`: Pub 2.10.5
- OS version: Windows 10 (2004)
- Are you using the Chinese community mirror or a corporate firewall? **Yes**

# Problem

Pub command exit unexpectedly once a pub command is finished, causing commands after a pub command unable to run. 

## Step to reproduce

1. Run `pub version` and `dart --version` separately in cmd.
2. Write the below content into `test.bat`.
```batch
pub version && dart --version
pub version
dart --version
pause
```
3. Execute `test.bat` in cmd.
4. Only the first line of commands has output.

![image](https://user-images.githubusercontent.com/15884415/107138170-c8f79880-694d-11eb-8a17-37f232313ed5.png)

## Expected behavior

All commands can be executed.

## Actual behavior

Commands after a valid pub command cannot be reached.

**Comment by CaiJingLong** [comment]
This unexpected exit will make it impossible to create an automatic build script under windows.

such as `build.bat`:

```cmd
pub get
:: The dart2native command cannot be run.
dart2native bin/main.dart
```

**Comment by jonasfj** [comment]
Any ideas why? Is it exiting non-zero when it shouldn't.

I don't know batch that well, or why it would stop running the script.

Maybe there is some interaction between the bat script wrapping dart.

FYI, `dart pub` has replaced `pub`, let's debug the issue for that command if it's still present there.

**Comment by AlexV525** [comment]
It's not only affects `pub` itself, but other snapshots that have been global activated. Such as `pub global activate flutter_asset_generator` then `fgen`, result the same issue. And only on Windows.

**Comment by AlexV525** [comment]
> Any ideas why? Is it exiting non-zero when it shouldn't.

Agreed at the first look, need more investigation.

> I don't know batch that well, or why it would stop running the script.

Maybe pub detects a line output in the standard console or something, then finish itself.

> Maybe there is some interaction between the bat script wrapping dart.

One more details about this is it's the same behavior if you run a snapshot in the IDE's console, such as Android Studio.

> FYI, `dart pub` has replaced `pub`, let's debug the issue for that command if it's still present there.

```batch
dart pub version
dart --version
```

Batch like this still failed on Windows.
![image](https://user-images.githubusercontent.com/15884415/107154926-7d2a0b00-69b0-11eb-8c04-bc8b41563964.png)

**Comment by AlexV525** [comment]
Some relative source, cannot tell whether they are the core code for this issue.

https://github.com/dart-lang/pub/blob/9a70949e25e9407423c68af5aed0f52b0655fe09/lib/src/io.dart#L759-L762

https://github.com/dart-lang/pub/blob/9a70949e25e9407423c68af5aed0f52b0655fe09/lib/src/io.dart#L564-L567

**Comment by AlexV525** [comment]
Tested on the master channel of flutter (since Idk how to compile the `pub` snapshot), with the batch below. The console will not close after a snapshot command executed, but further commands still unable to execute. Also the snapshot seems to be executed twice.

```batch
fgen -o lib\constants\resources.dart --no-preview --no-watch
dart --version
```

(fgen is installed from `dart pub global activate flutter_asset_generator`)

```shell
$ dart --version
Dart SDK version: 2.13.0-0.0.dev (dev) (Thu Feb 4 08:08:19 2021 -0800) on "windows_x64"
```

![image](https://user-images.githubusercontent.com/15884415/107182912-a1b9cd80-6a18-11eb-8482-5a19b88d7f8f.png)

**Comment by sigurdm** [comment]
This doesn't have anything to do with pub really - but specifics of batch files.
pub is a batch file (pub.bat). And if you execute a batch-file from a batch-file it will take over control (never return to the caller).

https://stackoverflow.com/a/4799002/4809566 explains the issue.

The solution is to use: ```call pub --version```

One could argue that pub in the sdk should be an exe. But I think it is too late to change that now without being breaking - and as jonasfj@ said, you should use `dart pub` instead!

**Comment by AlexV525** [comment]
Thanks for the explanation! But here I discovered that the snapshot has been executed twice. I'll file another issue later.

**Comment by AlexV525** [comment]
Thought about this again last night, it's gonna be a common pitfall for windows users that a snapshot actually executed with a batch file. They don't usually familiar with batch or pub itself. It's still worth to add a warning tip at pub's documentation.

**Comment by sigurdm** [comment]
@jonasfj WDYT? Should we warn about this?
I feel it is out of scope for dart/pub documentation - but I guess it could help someone from getting stuck.

**Comment by jonasfj** [comment]
I think `pub` is deprecated, in favor of `dart pub`.

For globally activated packages maybe we can consider something other than `.bat` scripts, or compiling executables, but that might be a bit more work.

**Comment by AlexV525** [comment]
Executables seems more reasonable, since `dart2native` is quite simple to do the compile job.

**Comment by JHubi1** [comment]
Thank god, this saved my day. To use a flutter command on Windows inside a bat file, you have to use `call flutter ...`, no way I would have found that out myself.

**Comment by sigurdm** [comment]
> no way I would have found that out myself.

Yeah that story really sucks - but it is not that easy to change.
